### PR TITLE
Replace truncating fp32->bf16 with tie-to-even rounding rule

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
Github Issues: 
https://github.com/tenstorrent/tt-metal/issues/19881 
https://github.com/tenstorrent/tt-metal/issues/16579

### Problem description
Conversion from fp32 to bf16 in host truncates last 23-7 bits, but to convert to same exact number as pytorch and wormhole does, it should be properly rounded with tie-to-even rule. 

Here is a difference example:
```
x = 1e-08
  Trunc round-trip = 9.95351e-09
  RN    round-trip = 1.00117e-08
x = 3.14
  Trunc round-trip = 3.125
  RN    round-trip = 3.14062
```

### What's changed
- I used tie-to-even rounding that was used in moreh_full before. It seems to exactly replicate pytorch's behavior.
- Improves overall training convergence in tt-train, but I needed to increase threshold for some tt-train tests. 
- Replace `assert_with_pcc` with `assert_equal` in most creation tests
- Leave truncation in scalar passing to kernel in binary ops. 

This is third time I try to merge this commit. 
- First PR: https://github.com/tenstorrent/tt-metal/pull/19882 
- First revert was due to broken +-inf/nan handling on device after tie-to-even rounding (revert: https://github.com/tenstorrent/tt-metal/pull/26098, fix commit: https://github.com/tenstorrent/tt-metal/pull/26123) 
- Second revert was due to stable diffusion accuracy degradation (https://github.com/tenstorrent/tt-metal/pull/26258)

To fix stable diffusion I had to replace tie-to-even with truncation always when passing scalars to binary ops. I don't know what is the root issue here, but this is a minimal change that fixes this test. 

### Checklist
- [X] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes -- https://github.com/tenstorrent/tt-metal/actions/runs/17234423743
- [X] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) -- https://github.com/tenstorrent/tt-metal/actions/runs/17234431304
- [X] (Single-card) Frequent model and ttnn tests -- https://github.com/tenstorrent/tt-metal/actions/runs/17234436541 
- [X] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) -- https://github.com/tenstorrent/tt-metal/actions/runs/17150738095
- [X] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) -- https://github.com/tenstorrent/tt-metal/actions/runs/17234440299
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models). -- https://github.com/tenstorrent/tt-metal/actions/runs/17150747536 (also fails on main)
- [X] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable) -- https://github.com/tenstorrent/tt-metal/actions/runs/17234440299
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work (fails on main)
- [X] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main) -- https://github.com/tenstorrent/tt-metal/actions/runs/17234445571
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release) -- -- https://github.com/tenstorrent/tt-metal/actions/runs/17150764016 (also fails on main)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by philei-tt** [comment]
(Single-card) Frequent model and ttnn tests: https://github.com/tenstorrent/tt-metal/actions/runs/17104571322

**Comment by philei-tt** [comment]
APC: https://github.com/tenstorrent/tt-metal/actions/runs/17104603948

**Comment by philei-tt** [comment]
After local fix of stable diffusion: 
APC: https://github.com/tenstorrent/tt-metal/actions/runs/17126734395 
Single-card) Frequent model and ttnn tests: https://github.com/tenstorrent/tt-metal/actions/runs/17126763010

File: ttnn/cpp/ttnn/operations/full/device/full_program_factory.cpp [hunk 1]
```diff
@@ -1,37 +1,16 @@
-// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include "full_device_operation.hpp"
 #include <tt-metalium/work_split.hpp>
+#include <tt-metalium/bfloat16.hpp>
+#include "full_device_operation.hpp"
```
Comment (nvelickovicTT) [inline ttnn/cpp/ttnn/operations/full/device/full_program_factory.cpp]: Ideally, this include shouldn't need to be moved. And if ti needs, it signals it might not be self sufficient.

**Review by nvelickovicTT — APPROVED** [review]

**Review by omilyutin-tt — APPROVED** [review]

**Review by dchenTT — APPROVED** [review]

File: tt_metal/impl/data_format/bfloat16.cpp [hunk 2]
```diff
@@ -13,6 +13,37 @@
...
+        union {
+            uint32_t U32;
+            float F32;
+        };
+
+        F32 = val;
+        // Rounding bias = 0111 1111 1111 1111 (0x7FFF) if last bit of mantissa ((U32 >> 16) & 1)
+        // is 0, otherwise 1000 0000 0000 0000 (0x8000).
+        // This ensures that we round to the nearest even number.
+        uint32_t rounding_bias = ((U32 >> 16) & 1) + UINT32_C(0x7FFF);
```
Comment (amahmudTT) [inline tt_metal/impl/data_format/bfloat16.cpp:27]: I am probably missing something totally. You said if the last bit of mantissa  ((U32 >> 16) & 1)
How is that the last bit of mantissa ? If I am not getting it , could you elaborate it in your comment so that the code is easily understandable.

Comment (amahmudTT) [reply tt_metal/impl/data_format/bfloat16.cpp:27]: And if the trick you are applying has a name or documented somewhere, please link it. Thank you.

**Review by dmakoviichuk-tt — APPROVED** [review]

File: tt_metal/impl/data_format/bfloat16.cpp (hunk 2 above)
Comment (philei-tt) [reply tt_metal/impl/data_format/bfloat16.cpp:27]: Last bit of mantissa for truncated bf16 representation (U32>>16)

Comment (philei-tt) [reply tt_metal/impl/data_format/bfloat16.cpp:27]: Not sure about trick documentation, since bf16 is not a part of IEEE754, and it specification and conversion spec is inferred from other types. 

Pretty much identical logic is used in 
- Eigen: https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/arch/Default/BFloat16.h?utm_source=chatgpt.com#L565 
- Pytorch: https://github.com/pytorch/pytorch/blob/e9d42b3880dcdbd823bbdc9370c8b0b3af0ba2e3/torch/headeronly/util/BFloat16.h#L98
- ROCM HIP: https://rocm.docs.amd.com/projects/HIP/en/docs-5.2.3/doxygen/html/hip__bfloat16_8h_source.html  
- CUDA: https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH____BFLOAT16__MISC.html?utm_source=chatgpt.com#_CPPv421__float22bfloat162_rnK6float2

Comment (philei-tt) [reply tt_metal/impl/data_format/bfloat16.cpp:27]: offtop, I just noticed that I can replace union with std::bitcast, which should be safer and more readable. UPD: updated

File: ttnn/cpp/ttnn/operations/full/device/full_program_factory.cpp (hunk 1 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/full/device/full_program_factory.cpp]: You're right, probably IDE messed up includes. Good catch

Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/full/device/full_program_factory.cpp]: Fixed

**Comment by philei-tt** [comment]
Rerunning everything again before merging: 

https://github.com/tenstorrent/tt-metal/actions/runs/17234423743
https://github.com/tenstorrent/tt-metal/actions/runs/17234431304
https://github.com/tenstorrent/tt-metal/actions/runs/17234436541
https://github.com/tenstorrent/tt-metal/actions/runs/17234440299
https://github.com/tenstorrent/tt-metal/actions/runs/17234443373
https://github.com/tenstorrent/tt-metal/actions/runs/17234445571

**Review by amahmudTT — APPROVED** [review]
